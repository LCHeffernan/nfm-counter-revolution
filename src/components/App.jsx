import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import "../styles/App.css";
import Content from "./Content";
import DiscController from "./DiscController";

const App = ({ figures }) => {
  const baseAnimationDurationMs = 1200;
  const appIdleTimerDurationMillis = 5000;
  const [index, setIndex] = useState(0);
  const [isAppIdle, setIsAppIdle] = useState(true);
  const [appIdleTimer, setAppIdleTimer] = useState();

  const incrementIndex = () =>
    setIndex((prev) => (prev < figures.length - 1 ? prev + 1 : 0));

  const startAppIdleTimer = () => {
    setIsAppIdle(false);
    setAppIdleTimer(clearTimeout(appIdleTimer));
  };

  useEffect(() => {
    const doc = document.documentElement;
    doc.style.setProperty(
      "--base-animation-duration",
      `${baseAnimationDurationMs}ms`
    );
  }, []);

  useEffect(() => {
    if (!(isAppIdle || appIdleTimer)) {
      setAppIdleTimer(
        setTimeout(() => {
          setIsAppIdle(true);
          setAppIdleTimer(clearTimeout(appIdleTimer));
        }, appIdleTimerDurationMillis)
      );
    }
    return () => clearTimeout(appIdleTimer);
  }, [isAppIdle, appIdleTimer]);

  return (
    <div className="App">
      <main>
        <DiscController
          incrementIndex={incrementIndex}
          setIndex={setIndex}
          index={index}
          figures={figures}
          baseAnimationDurationMs={baseAnimationDurationMs}
          startAppIdleTimer={startAppIdleTimer}
          isAppIdle={isAppIdle}
        />
        <Content
          figures={figures}
          index={index}
          baseAnimationDurationMs={baseAnimationDurationMs}
          isAppIdle={isAppIdle}
        />
      </main>
    </div>
  );
};

App.propTypes = {
  figures: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number,
      title: PropTypes.string,
      angle: PropTypes.number,
      text: PropTypes.arrayOf(PropTypes.string),
    })
  ).isRequired,
};

export default App;
