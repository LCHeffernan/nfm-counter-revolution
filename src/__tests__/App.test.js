import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import App from "../components/App";

const validProps = {
  figures: [
    {
      id: 0,
      title: "figure 0",
      angle: 0,
      text: [
        "Figure 0 Paragraph 1",
        "Figure 0 Paragraph 2",
        "Figure 0 Paragraph 3",
      ],
    },
    {
      id: 1,
      title: "figure 1",
      angle: -60,
      text: [
        "Figure 1 Paragraph 1",
        "Figure 1 Paragraph 2",
        "Figure 1 Paragraph 3",
      ],
    },
    {
      id: 2,
      title: "figure 2",
      angle: -120,
      text: [
        "Figure 2 Paragraph 1",
        "Figure 2 Paragraph 2",
        "Figure 2 Paragraph 3",
      ],
    },
  ],
};
const renderApp = () => render(<App {...validProps} />);

describe("App", () => {
  it("matches snapshot", () => {
    const { asFragment } = renderApp();

    expect(asFragment()).toMatchSnapshot();
  });

  beforeEach(() => renderApp());

  it("renders the Content component", () => {
    expect(screen.getByTestId("content")).toBeInTheDocument();
  });

  it("renders the DiscController component", () => {
    expect(screen.getByTestId("disc-controller")).toBeInTheDocument();
  });
});

describe("DiscController component", () => {
  beforeEach(() => renderApp());

  it("button correctly updates Content component text content", () => {
    const button = screen.getByRole("button");

    validProps.figures.forEach((figure, i) => {
      expect(screen.queryByText(figure.text[0])).toBeInTheDocument();

      fireEvent.click(button);

      if (i === validProps.figures.length - 1) {
        expect(
          screen.queryByText(validProps.figures[0].toHaveClass("text-fade-in"))
        ).toBeInTheDocument();
      } else {
        expect(screen.queryByText(figure.text[0])).toEqual(null);
      }
    });
  });
});
