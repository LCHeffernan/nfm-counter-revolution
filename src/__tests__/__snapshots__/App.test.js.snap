// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <header>
      <h1>
        figure 0
      </h1>
    </header>
    <main>
      <div
        class="content"
        data-testid="content"
      >
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Quisque sit amet hendrerit augue, quis commodo quam. Morbi non justo ut odio dictum consectetur. Vivamus dignissim metus nec ligula laoreet, ut auctor nunc rutrum. Mauris vitae nulla quam. Morbi sed sem nulla. Nulla facilisi. Nam faucibus accumsan iaculis. Suspendisse ac dapibus mi. Sed eget velit mi. Nullam blandit, leo at faucibus feugiat, leo eros viverra turpis, vel finibus mi augue et eros.Nunc nisl lacus, mollis ut fermentum et, tristique id odio. Nam tincidunt bibendum convallis. Donec laoreet sapien eu tortor tristique auctor. Donec ullamcorper malesuada lorem, eu sodales erat ornare ac. Etiam nec mauris sit amet velit molestie posuere sit amet nec elit. Curabitur id massa non libero ullamcorper porta. Phasellus sed leo eu nisi vehicula ornare. Integer rhoncus venenatis odio, at placerat turpis consectetur id. Fusce sodales accumsan feugiat. Quisque fermentum malesuada pharetra. Vestibulum dui lacus, efficitur sit amet est eget, efficitur elementum urna. Aliquam sodales a augue sit amet malesuada. Nulla semper vitae libero at semper. In sit amet posuere tortorNulla facilisi. Donec et egestas erat. In sed aliquam leo. Aenean eu mi eget libero sodales viverra eget nec tellus. Fusce egestas magna purus, nec vehicula sapien.
      </div>
      <div
        class="revolver"
        data-testid="revolver"
      >
        <button
          type="button"
        >
          increment index
        </button>
      </div>
    </main>
  </div>
</DocumentFragment>
`;
